bondcpp/include/bondcpp/bond.hpp

```diff
@@ -163,50 +194,50 @@ class Bond
   void doPublishing();
   void publishStatus(bool active);
 
-  std::vector<std::function<void(void)>> pending_callbacks_;
   void flushPendingCallbacks();
 
   rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base_;
-  rclcpp::node_interfaces::NodeTimersInterface::SharedPtr node_timers_;
   rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging_;
+  rclcpp::node_interfaces::NodeTimersInterface::SharedPtr node_timers_;
 
   rclcpp::TimerBase::SharedPtr connect_timer_;
   rclcpp::TimerBase::SharedPtr disconnect_timer_;
   rclcpp::TimerBase::SharedPtr heartbeat_timer_;
   rclcpp::TimerBase::SharedPtr publishing_timer_;
   rclcpp::TimerBase::SharedPtr deadpublishing_timer_;
 
+  std::mutex state_machine_mutex_;
   std::unique_ptr<BondSM> bondsm_;
   BondSMContext sm_;
 
   std::string topic_;
   std::string id_;
   std::string instance_id_;
   std::string sister_instance_id_;
-  std::function<void(void)> on_broken_;
-  std::function<void(void)> on_formed_;
-  bool sisterDiedFirst_;
-  bool started_;
-  bool connect_timer_reset_flag_;
-  bool disconnect_timer_reset_flag_;
-  bool deadpublishing_timer_reset_flag_;
-  bool disable_heartbeat_timeout_;
-  std::mutex mutex_;
-  std::condition_variable condition_;
-
-  double connect_timeout_;
-  double heartbeat_timeout_;
-  double disconnect_timeout_;
-  double heartbeat_period_;
-  double dead_publish_period_;
+
+  std::mutex callbacks_mutex_;
+  std::vector<EventCallback> pending_callbacks_;
+  EventCallback on_broken_;
+  EventCallback on_formed_;
+
+  bool sisterDiedFirst_ {false};
+  bool started_ {false};
+  bool connect_timer_reset_flag_ {false};
+  bool disconnect_timer_reset_flag_ {false};
+  bool deadpublishing_timer_reset_flag_ {false};
+  bool disable_heartbeat_timeout_ {false};
+
+  rclcpp::Duration connect_timeout_;
+  rclcpp::Duration disconnect_timeout_;
+  rclcpp::Duration heartbeat_timeout_;
+  rclcpp::Duration heartbeat_period_;
+  rclcpp::Duration dead_publish_period_;
 
   rclcpp::Subscription<bond::msg::Status>::SharedPtr sub_;
   rclcpp::Publisher<bond::msg::Status>::SharedPtr pub_;
 };
-
 }  // namespace bond
 
-
 // Internal use only
 struct BondSM
```

Comment:
Since this is internal use only, can we instead make this private within the `Bond` class?  That would also remove the need to make it a `friend` above.

(though this is starting to drift further away from the initial goal of this PR, so feel free to punt on this)

Reply:
I played with this for a second, and the interactions with the generated state machine code are bad.  I'm going to punt on this for now.

Reply:
Yep, that's fine.  Thanks for taking a look.